main/autogen/subclasses.cc

```diff
@@ -92,34 +84,57 @@ optional<Subclasses::SubclassInfo> Subclasses::descendantsOf(const Subclasses::M
+    for (auto &n : absl::StrSplit(rawName, "::")) {
+        const auto &nameRef = gs.lookupNameUTF8(n);
+        if (!nameRef.exists())
+            return core::Symbols::noSymbol();
+        sym = gs.lookupClassSymbol(sym, gs.lookupNameConstant(nameRef));
+    }
+    return sym;
+}
+
 // Manually patch the child map to account for inheritance that happens at runtime `self.included`
 // Please do not add to this list.
-void Subclasses::patchChildMap(Subclasses::Map &childMap) {
-    childMap["Opus::SafeMachine"].entries.insert(childMap["Opus::Risk::Model::Mixins::RiskSafeMachine"].entries.begin(),
-                                                 childMap["Opus::Risk::Model::Mixins::RiskSafeMachine"].entries.end());
+void Subclasses::patchChildMap(const core::GlobalState &gs, Subclasses::Map &childMap) {
+    auto safeMachineRef = getConstantRef(gs, "Opus::SafeMachine");
+    auto riskSafeMachineRef = getConstantRef(gs, "Opus::Risk::Model::Mixins::RiskSafeMachine");
+    if (!safeMachineRef.exists() || !riskSafeMachineRef.exists())
+        return;
+    childMap[safeMachineRef].entries.insert(childMap[riskSafeMachineRef].entries.begin(),
+                                            childMap[riskSafeMachineRef].entries.end());
 }
 
-vector<string> Subclasses::serializeSubclassMap(const Subclasses::Map &descendantsMap,
-                                                const vector<string> &parentNames) {
+vector<string> Subclasses::serializeSubclassMap(const core::GlobalState &gs, const Subclasses::Map &descendantsMap,
+                                                const vector<core::SymbolRef> &parentNames, const bool showPaths) {
     vector<string> descendantsMapSerialized;
+    const auto classFormatString = showPaths ? " class {} {}" : " class {}";
+    const auto moduleFormatString = showPaths ? " module {} {}" : " module {}";
 
-    for (const string &parentName : parentNames) {
-        auto fnd = descendantsMap.find(parentName);
+    for (const auto &parentRef : parentNames) {
+        auto fnd = descendantsMap.find(parentRef);
         if (fnd == descendantsMap.end()) {
             continue;
         }
         const Subclasses::SubclassInfo &children = fnd->second;
 
+        string parentName = parentRef.show(gs);
+
         auto type = children.classKind == ClassKind::Class ? "class" : "module";
         descendantsMapSerialized.emplace_back(fmt::format("{} {}", type, parentName));
 
         auto subclassesStart = descendantsMapSerialized.size();
-        for (const auto &[name, type] : children.entries) {
+        for (const auto &sym : children.entries) {
+            string_view path = gs.getPrintablePath(sym.loc(gs).file().data(gs).path());
+            string childName = sym.show(gs);
             // Ignore Modules
-            if (type == autogen::Definition::Type::Class) {
-                descendantsMapSerialized.emplace_back(fmt::format(" class {}", name));
+            if (sym.asClassOrModuleRef().data(gs)->isClass()) {
```

Comment:
Seeing this `asClassOrModuleRef` suggests to me that we wish `children.entries` had type `UnorderedSet<core::ClassOrModuleRef>` instead of `UnorderedSet<core::SymbolRef>`.

There are a handful of symbols that are constants but not classes/modules: static constant assignments like `X = 1` and type members like `X = type_member`.

It seems that right now, the code depends on `parent_of.exists()` implying that `parent_of.data(pf).sym.isClassOrModule()`. I think that's a safe assumption?

In that case, I'd prefer to move the invariant closer to construction, namely moving this `asClassOrModuleRef` to `listAllSubClasses`.

Reply:
<img width="865" alt="Screenshot 2023-08-09 at 4 06 31 PM" src="https://github.com/sorbet/sorbet/assets/5544532/a4c94b6e-7c8d-4a98-b9b9-28b21c9d66c9">

Reply:
ooh yes, that's even better.